### PR TITLE
feat: generator detects diregapic repeated/map field for paged methods

### PR DESCRIPTION
**Pull request by software-dov**
Certain APIs follow a different convention than AIP-4233 for paginated
methods:
* They permit either a page_size or max_results field with identical
semantics
* They allow pagination over maps.

The existing paging detection logic now explicitly checks if the
repeated field is mapped and fails if it is.
New paging detection logic checks for the updated semantics for use in
future generated code changes.

### CODE REVIEW

File: internal/gengapic/paging.go
Comment (vchudnov-g) [reply internal/gengapic/paging.go:177]: I would consider changing the name to something like `getPagingFields` because:
- plural: you're getting more than one field
- diregapic: this returns fields for either the diregapic or non-diregapic case; ie it handles both scenarios

File: internal/gengapic/paging_test.go
Comment (vchudnov-g) [inline internal/gengapic/paging_test.go:473]: nit: "DIREGAPIC" capitalization?

File: internal/gengapic/paging_test.go
Comment (vchudnov-g) [inline internal/gengapic/paging_test.go:472]: The `diregapicPagingField` function will return non-nils even  for correct but non-DIREGAPIC paging, IIUC?

File: internal/gengapic/paging.go
Comment (vchudnov-g) [inline internal/gengapic/paging.go:141]: It seems like `diregapicPagingField()` covers a superset of functionality as `pagingField()` and similarly for tests, so there's some duplicated logic. Would it make sense to simply have one function with a flag `allowDIREGAPIC bool`?

File: internal/gengapic/paging.go
Comment (vchudnov-g) [reply internal/gengapic/paging.go:177]: (or, in the spirit of the previous comment, making this essentially be _the_ `pagingFields()` function.

Comment (software-dov) [reply internal/gengapic/paging.go:177]: Good point; named returns sound like the correct thing here.
You're right: the diregapic paging fields are a superset of the 'regular' paging fields, so it might make sense to use the same function for both. The one exception I can think of is if the response has multiple repeated fields. I'll think about that some more.

File: internal/gengapic/paging_test.go
Comment (software-dov) [reply internal/gengapic/paging_test.go:472]: Correct. My original thought was to check for regular paging first then diregapic paging. Since it's a superset, however, it makes sense to just have the one function.

File: internal/gengapic/paging.go
Comment (software-dov) [reply internal/gengapic/paging.go:141]: I think it would make sense to do that. This brings us back to the question of having an explicit diregapic flag, though. I'm in favor of not having a flag and instead treating it like diregapic pagination if it passes the predicate. The two differences are the name variant of page size field, i.e. `max_results`, and maps.

**Comment by software-dov** [comment]
Integrated Victor's proposals. There is now a single `getPagingFields` method in the generator that does not differentiate between regular methods and diregapic methods.

**Comment by noahdietz** [comment]
There are still a few open questions on the design for this. Can we nail those down before submitting code?

File: internal/gengapic/paging.go
Comment (vchudnov-g) [inline internal/gengapic/paging.go:155]: How about:
```
if pagedField == nil || hasNextPageToken == nil {
  // not paging
  return nil, nil, nil
}
return pagedField, pageSizeField, nil
```

File: internal/gengapic/paging.go
Comment (vchudnov-g) [inline internal/gengapic/paging.go:152]: Where do you use `hasNextPageToken`? I suggest something in my other comment below.

File: internal/gengapic/paging.go
Comment (vchudnov-g) [inline internal/gengapic/paging.go:137]: You're not checking for map responses yet? I would either
(a) add a TODO and change the PR description to say you're deferring map-method detection or
(b) detect map responses here and now and add an error/TODO in the paging iterator for map responses

File: internal/gengapic/paging.go
Comment (vchudnov-g) [reply internal/gengapic/paging.go]: Well, the function could have a parameter `allowDIREGAPIC bool`, and we can always set it to `true` at calling sites for now. If/when we introduce a flag for DIREGAPIC, that value will be taken from the flag. 

- Advantage: it marks and flags the DIREGAPIC-specific code right now, while it's fresh in our mind. I t also allows us to test the code with the parameter set to `false` right away.

- Disadvantage: until we do have a top-level flag, all the calling sites will have an extra parameter `(.., true)` essentially. (Though we should make it into a constant for comprehensibility: `(..., allowDIREGAPIC)`

Actually, here's an idea to get rid of that disadvantage: have `allowDIREGAPIC` be a (package?) var referenced *inside* the function(s) that have DIREGAPIC-specific code, and NOT a function param. That way we can still flag DIREGAPIC-specific sections of code, and because it's a test, we can set it to false and test non-DIREGAPIC flows (careful with parallel tests!). It only becomes a function param instead of a package var when we do have a DIREGAPIC flag to pass.

WDYT?

**Comment by vchudnov-g** [comment]
Also, could you have a companion PR so we can see what the generated code is like? That is always very helpful. Thanks!

**Review by noahdietz — COMMENTED** [review]
If I am understanding, this should have 0 affect on existing clients, correct? Thank you for breaking up the internal API changes and the feature work.

File: internal/gengapic/paging.go
Comment (noahdietz) [reply internal/gengapic/paging.go]: > have allowDIREGAPIC be a (package?) var referenced inside the function(s) that have DIREGAPIC-specific code

Maybe it would be cleaner to add a new field in the `generator.options` struct that we use to carry both flags and other config at the generator-level. And if there is ever a plugin option added, it just gets wired up to that.

Comment (vchudnov-g) [reply internal/gengapic/paging.go]: +1, even better! This will allow the tests to be modular without much worry, while preserving all the other benefits.

File: internal/gengapic/paging.go
Comment (software-dov) [reply internal/gengapic/paging.go:107]: You are correct, it should not be. It slipped through my fingers as I pithed the method.

File: internal/gengapic/paging.go
Comment (software-dov) [reply internal/gengapic/paging.go:105]: It is changed because this method was originally separate instead of replacing `pagingField` and had different errors but it subsumed `pagingField`.

File: internal/gengapic/paging.go
Comment (software-dov) [reply internal/gengapic/paging.go:137]: Before checking the options struct for diregapic mode, there's not a lot to do for detecting map responses here. They are also repeated; the main difference would be that we need to generate a custom pair type and use that for the iterator.

File: internal/gengapic/paging.go
Comment (software-dov) [reply internal/gengapic/paging.go:152]: Good catch. This is a test lacuna.

File: internal/gengapic/paging.go
Comment (software-dov) [reply internal/gengapic/paging.go:139]: Yes, that's exactly what I meant :P
I am not sure if it's guaranteed that the order is preserved. The previous code certainly acted as if it was preserved, and the wording for AIP-4233 strongly implies that it is.

File: internal/gengapic/paging_test.go
Comment (software-dov) [reply internal/gengapic/paging_test.go]: Just unilaterally removed the diregapic moniker from the code.

File: internal/gengapic/paging.go
Comment (software-dov) [reply internal/gengapic/paging.go]: Not dropping the diregapic option, just holding off on this for a little bit. Will post code containing it soon.

Comment (vchudnov-g) [reply internal/gengapic/paging.go]: OK. I think having the bool be always be set in the struct for now make sense, and having code here and in tests use it. And like Noah says, we can wire up to any eventual plugin option later.

File: internal/gengapic/paging.go
Comment (noahdietz) [inline internal/gengapic/paging.go:58]: Just to make sure, we won't really ever get to this code path today with `compute` because `getPagingFields` would say "this is not a paginated rpc", right? I'm just concerned with this breaking `compute` generation today.

Comment (vchudnov-g) [reply internal/gengapic/paging.go:58]: I gather from the tests  that a map field is represented as a repeated set of (key,value) tuples, so in that case the old version of `pagingFields()` would have recognized this as a paging field (assuming `page_size`, `page_token`, and `next_page_token` were there) and this iterator would have done something? Am I interpreting correctly?

That said, we're in preview, so we do have the ability to make breaking changes right now.

**Review by vchudnov-g — APPROVED** [review]
I can't vouch for the representation of mapped fields in protos, but otherwise LGTM

File: internal/gengapic/paging.go
Comment (noahdietz) [reply internal/gengapic/paging.go:58]: I'm not concerned with breaking changes to the client surface, I'm concerned with failure during generation, breaking the build.

Comment (vchudnov-g) [reply internal/gengapic/paging.go:58]: Oh, OK.

@software-dov it would still be useful to have a companion PR with the generated code; it's easier top spot some issues there. And to Noah's point, were you able to run CI test locally to verify no breakages should occur?

Comment (noahdietz) [reply internal/gengapic/paging.go:58]: We don't have a `map<string, Message>` in Showcase at the moment (you were working on that), so this specific case wouldn't be exercised I believe. A chicken in the egg problem. I would say instead of returning a generator error here, add code to `getPagingFields` that skips those RPCs that only have a `map` for pagination, treating them as normal Unary RPCs. This will of course cause a breaking change when we implement mapped pagination if we were to release this skipping behavior. Not sure this is a big deal though given the Compute client is in alpha for us.

Comment (noahdietz) [reply internal/gengapic/paging.go:58]: [This is the hacky example](https://github.com/googleapis/gapic-generator-go/pull/716) of code that would skip mapped pagination.

Comment (vchudnov-g) [reply internal/gengapic/paging.go:58]: True, yes. Probably the best thing is to skip mapped pagination for right now (that will unblock my upcoming Showcase changes) and then implement it shortly after that.

My very slight inclination is to not do a new release until we have both max_results and mapped_pagination in, but I'm fine doing a release without mapped pagination and then issuing a breaking change with it. After all, we are still alpha and have forewarned users.

Comment (noahdietz) [reply internal/gengapic/paging.go:58]: > not do a new release until we have both max_results and mapped_pagination in

sgtm! But one reason we'd want to is that Showcase regen CI uses the latest released version of the generator. So without a released version that handles/skips mapped pagination, your changes will still fail. That's the chicken and the egg problem here.

Comment (software-dov) [reply internal/gengapic/paging.go:58]: Fine by me. I'll add an escape hatch for mapped pagination and tweak the tests real quick.

**Review by noahdietz — APPROVED** [review]

**Pull request merged**
